src/software/backend/simulation/physics_simulator.cpp

```diff
@@ -0,0 +1,40 @@
+#include "software/backend/simulation/physics_simulator.h"
+
+#include <Box2D/Box2D.h>
+
+PhysicsSimulator::PhysicsSimulator(const World& world)
+{
+    b2Vec2 gravity(0, 0);
+    physics_world           = std::make_shared<b2World>(gravity);
+    physics_world_timestamp = Timestamp::fromSeconds(0);
+
+    setWorld(world);
+}
+
+World PhysicsSimulator::stepSimulation(const Duration& time_step)
+{
+    physics_world->Step(time_step.getSeconds(), velocity_iterations, position_iterations);
+    physics_world_timestamp = physics_world_timestamp + time_step;
+
+    World new_world;
+    new_world.updateTimestamp(physics_world_timestamp);
+    if (physics_ball)
+    {
+        new_world.mutableBall() =
+            physics_ball->getBallWithTimestamp(physics_world_timestamp);
+    }
+    if (physics_field)
+    {
+        new_world.mutableField() =
+            physics_field->getFieldWithTimestamp(physics_world_timestamp);
+    }
```

Comment:
I think having a PhysicsWorld abstraction might help here as we can have it contain the PhysicsBall and PhysicsField instead of the Simulator containing them. It could also make scaling easier. For example, if we decide to add PhysicsRobots and such.

Reply:
Yes I still need to add that, as well as `PhysicsRobot` and other abstractions. I had enough to start figuring out how everything would fit together at a high level so did that first.

#768 shows the breakout of the issues I still need to do

Reply:
I second this.